### ISSUE TITLE
Change Request: Modifications to Admin API to support Settlement

### ISSUE THREAD

**Issue by MichaelJBRichards** [mojaloop/mojaloop-specification#117]
### **Table of Contents**

* [1. Preface](#1.-preface)
  * [1.1 Change Request Information](#1.1-change-request-information)
  * [1.2 Document Version Information](#1.2-document-version-information)
* [2. Problem Description](#2.-problem-description)
  * [2.1 Background](#2.1-background)
  * [2.2 Current Behaviour](#2.2-current-behaviour)
  * [2.3 Requested Behaviour](#2.3-requested-behaviour)
* [3 Proposed Solution Options](#3.-proposed-solution-options)

<h2 id="1.-preface">1. Preface</h2>
___

This section contains basic information regarding the change request.

<h4 id="1.1-change-request-information">1.1 Change Request Information</h4>

| Requested By | Michael Richards, INFITX |
| Change Request Status | In review ☒  / Approved ☐ / Rejected ☐ |
| Approved/Rejected Date |   |

<h4 id="1.2-document-version-information">1.2 Document Version Information</h4>

| Version | Date | Author | Change Description |
| --- | --- | --- | --- |
| 1.0 | 2023-02-07 | Michael Richards | Initial version of request. Sent out for review. |

<h2 id="2.-problem-description">2. Problem Description</h2>
___

<h4 id="2.1-background">2.1 Background</h4>

We are currently engaged in making changes to the way in which transfers are assigned for settlement in Mojaloop. A consequence of this is that transfers will be assigned deterministically to a settlement batch (we no longer refer to settlement windows) based on the characteristics of the transfer and the settlement model which is defined for transfers with those characteristics. Each settlement batch will be assigned a name which will reflect the characteristics of the batch. These names will include the following elements:

1. The name of the settlement model to which the batch belongs.
2. The time segment of the settlement model to which the batch belongs.
3. Any other characteristics defined for the settlement model.

The first two elements will be required; the third element may or may not be present.

<h4 id="2.2-current-behaviour">2.2 Current Behaviour</h4>

At present, an administrator requests initiation of a settlement by specifying the ID numbers of the settlement windows which are to be included in the settlement.

<h4 id="2.3-requested-behaviour">2.3 Requested Behaviour</h4>

Specification of the transfers to be included in a settlement should be made in the following way:

1. The administrator should always specify which settlement model is to be settled. This should be made from a list of the settlement models defined in the scheme, and it must therefore be possible to request a list of the settlement models defined in the scheme.
2. The administrator should specify what time period is to be settled, by specifying a time unit (e.g. days, hours, minutes) and a start and end point for the settlement expressed in those units. It should not be possible to specify a time period for settlement which is more granular than the time period specified for the settlement model.
3. If further segmentation of the settlement batches is specified by the settlement mode ad described in point 3 above, then the administrator should be able to request a list of the available segments for the settlement model and time period selected, and should be able to select one or more of the segments returned.

The administrator should also be able to select a group of batches previously selected by entering the ID number associated with the set of batches when they were originally returned.

Once an administrator has selected a group of batches using the method described above, they should be able to see the overall sum of the transfers in the batches selected, segmented by participant and optionally by batch. 

The administrator should also be able to see the status (open, closed, settling, settled) of the batches selected.

The set of batches selected should be assigned an ID. The administrator

If all of the batches selected are closed, then the administrator should be able to request that those batches are settled. The system should settle the batches selected and report the status of the settlement once it is complete. This report should be a synchronous response to the settlement request.

In addition, the admnistrator should be able to select a single settlement batch using the techniques described above and view the transfers which comprise it.

In addition, an administrator should be able to close a settlement batch manually. The single settlement batch should be selected either using the techniques described above, or by specifying a transfer ID and asking to close the settlement batch which contains that ID. 

<h2 id="3.-proposed-solution-options">3. Proposed Solution Options</h2>
___

**Comment by mdebarros** [comment]
## Action items from CCB Admin Meeting Held on 2023-02-08
- @pedrosousabarreto to create a PR to the Reference Architecture repo for the changes as discussed in the meeting
- @koekiebox add a comment here with an overview of the current API operatons for the vNext Settlements BC, with their descriptions, and a high-level process flow (_feel free to link/reference any existing documentation that you have developed for the detail_)

**Comment by pedrosousabarreto** [comment]
I've updated the reference architecture flow as discussed, here is the link that points to the updated settlements flow: https://miro.com/app/board/uXjVPpKjqlw=/?moveToWidget=3458764545671714541&cot=14

**Comment by koekiebox** [comment]
@mdebarros I have added section `#5` into the settlement `README.md` to list the current exposed API endpoints.
<img width="1037" alt="image" src="https://user-images.githubusercontent.com/1992108/225251764-6de739b8-2c14-40fa-9277-51fb1b638e8b.png">

https://github.com/mojaloop/settlements-bc/blob/main/docs/flows/README.md

**Comment by pedrosousabarreto** [comment]
Given that the normal flow is that settlements are triggered by the **TransferCommittedFulfiledEvt** message, that endpoint #01 should used for tests only, and maybe not part of the official API

**Comment by MichaelJBRichards** [comment]
My comments:
- As we've said, resource 1 is not relevant for the external API.
- For the external API, the POST /matrix request should return the current content of the matrix (i.e. the same as would be returned by resource 9.)
- As we've discussed, endpoint 3 is not required for the external API. Resource 9 should always return the current state of the matrix.
- On 4, we need to be able to close a batch by criteria without creating a matrix; and we need to be able to close the batch which contains a specified transfer, identified by transaction ID.
- We need an endpoint which enables an administrator to execute a settlement on a specified matrix. Which means: all open batches in the matrix should be closed. If there are any batches in the matrix which have already been settled, they should be excluded.
- Following on from the previous point, I think we probably need to segment the totals returned by resource 9. If the matrix contains batches which have already been settled, we would need to show: the total of settled batches, the total of unsettled batches (either open or closed) and the overall matrix total.
- I'm not sure how useful 6 and 7 are in an operational system. I would expect them to return far too much information for an operator to make sense of; and in the event that an operator does require to see every batch or account, they can do that simply by setting a wide set of criteria in resource 2.
- Surely resource 8 must be by batch, not every transfer in the settlement system? So an operator picks a batch, and says "show me what's in it".
In general: do we have content defined for the bodies of the endpoints? If so, can we discuss this too?

**Comment by pedrosousabarreto** [comment]
Given the comment, I'll also add my comments from the meeting of 15 Mar 2023.

Creation of a matrix is asynchronous as we cannot guarantee it will processed within HTTP timeout limits, hence #02 (POST matrix) must be asynchronous and return only a 202 status with the ID of the matrix to be created. Same pattern as the rest of our APIs that use the asynchronous pattern.

Request #03 is absolutely required, as a GET is cacheable by definition and cannot change the state of the system, so we need a separate method to request the recalculation of the matrix. In case of doubt see:  https://httpwg.org/specs/rfc9110.html#GET

According to the agreement of the meeting of the 8th Feb 2023, the closing of batches is done through the closing of a matrix that includes said batches, AND not the closing of the batches directly.

See note in updated ref arch miro board with the agreement of that meeting: https://miro.com/app/board/uXjVPpKjqlw=/?moveToWidget=3458764545671714541&cot=14

As discussed, a new method like 04 (POST /matrix/:id/close) will be created to settle (close&settle if target matrix is not closed).
With that, the operator has the ability to either close an open matrix, close&settle an open matrix or settle a closed matrix. This can be used to create a closed but not settled matrix which hold in dispute batches.

**Comment by MichaelJBRichards** [comment]
Don't care whether it's synchronous or not: in my opinion, an operator is entitled to expect a response from the "create a matrix" instruction.

I quote from the [URL ](https://httpwg.org/specs/rfc9110.html#GET)you refer: "intended to reflect the resource's current state". It is not intended to change the state of the system, merely to reflect its current state. I fail to see the relevance of the [second reference](https://developer.mozilla.org/en-US/docs/Web/HTTP/Methods/GET).

There may be a hair to be split relating to the difference between closing a batch and closing a single-batch matrix. I don't think we want to require the creation of a matrix in order to close a batch; and, as far as I can see, you omit consideration of the "close a batch which contains..." case.

We definitely need to be able to close an individual batch in case of dispute, in my opinion.

Sorry, I don't seem to be able to sign into the Miro board you link to. Please be more explicit.

I'm afraid I don't understand the last paragraph. Could you expand?

**Comment by mdebarros** [comment]
> Given that the normal flow is that settlements are triggered by the **TransferCommittedFulfiledEvt** message, that endpoint #1 should used for tests only, and maybe not part of the official API

In general, we have seperated out "test" APIs from the actual main API to the point that the "test" APIs on a service will run on a complete seperate port. 

Perhaps that is not necessary with RBAC, but I would still prefer to see that clear seperation.

**Comment by pedrosousabarreto** [comment]
@MichaelJBRichards The operator will get a response; that the request to create a matrix was accepted. It is up to the UI to redirect the operator to the page where the new matrix can be seen (with a get).
Looks like we are agreeing that a GET should not change the state of a matrix, but simply return it as it is persisted, without triggering any recalculation.

The last sentence was about your ask to have matrices (and included batches) that can be closed but not settled, I'll try to clarify better in an additional comment.

Agree @mdebarros, that post transfer request should be removed, it was there for the initial tests, and is no longer needed.

**Comment by koekiebox** [comment]
Included below (`DRAFT`) is the sub-list of endpoints for the Admin API. @MichaelJBRichards , valid point with regards to the endpoint bodies. There are bodies defined for settlement, but I think we should follow the same path as before. 

1. @koekiebox  would update the bodies for settlement to reflect the latest iteration - Will add here; https://github.com/mojaloop/settlements-bc/blob/main/docs/flows/README.md
2. We then have a discussion on the content of the bodies for each of the below endpoints for our next Admin API discussion?

| Ref # | URL                      | Method | Description                                                                                                                                                                                                                                                                                                                                         | 
|-------|--------------------------|--------|-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| `01.` | **/matrix**              | `POST` | Create a settlement matrix and return the matrix UUID *(newly generated)*. Batches that match the matrix criteria will be included as part of the settlement matrix. Any newly created batches will not be automatically associated with the settlement matrix. Newly created transfers will still be allocated to batches that are not yet closed. |
| `02.` | **/matrix/:id/close**    | `POST` | Close settlement batche/s. Once a settlement batch is closed, the balances for the batches and accounts will not change. No newly created transfers will be allocated to closed batches. Transaction ID's may be provided as parameters, in order to close batches containing those transfers.                                                      |
| `03.` | **/matrix/:id/settle**   | `POST` | Close settlement batche/s and settlement matrix. Generate a settlement matrix as response, as well as settling the batches. As part of the `settle` process, the open settlement batches will be closed. See `close`. An indicator will be present per batch to indicate whether the batch was closed as part of the `settle` action.               |
| `04.` | **/batches/:id**         | `GET`  | Retrieve settlement batch by UUID. The batches include the settlement accounts and balances.                                                                                                                                                                                                                                                        |
| `05.` | **/transfers**           | `GET`  | Retrieve all settlement transfers for a given batch.                                                                                                                                                                                                                                                                                                |
| `06.` | **/matrix/:id**          | `GET`  | Retrieve the settlement matrix by UUID. If the settlement matrix is not in a closed state, the batch and account balances for the open batches may change due to new transfers.                                                                                                                                                                     |

**Comment by MichaelJBRichards** [comment]
The problem here is that we've said that we want to define matrices by criteria, not by specifying individual batches. I don't see anything wrong with specifying batches by ID and closing them. They could still belong to the matrix they did before, except that settling that matrix while the batch is still in dispute would exclude the batch (I assume, by the way, that we have no reason not to allow a batch to belong to more than one matrix.) The matrix could then be settled again once the batch has been returned to a settleable state. I think that it's quite unlikely that we will want to close a set of batches because of a dispute, since this will generally be a consequence of a dispute about one (or perhaps a small number) of transfers. However, I wouldn't be against adding a resource to the **/matrix** endpoint to mark all batches in the matrix as being in dispute.

Presumably we shall also need an endpoint which enables an administrator to mark one or more batches as no longer being in dispute?

In response to your question, @koekiebox: we're expecting that a query might be raised about an individual transfer (or, as I say, a small number of transfers), and we would want to identify the batch to which it belongs and mark that batch as being in dispute, leaving subsequent transfers that would have been in that batch to go into a new sequence. That's why I've been asking consistently for an endpoint that enables an administrator to ask for the batch that contains a particular transfer (specified by transaction ID) to be marked as in dispute, which I think 4 does. So thanks for that.

On 6, presumably we would include the batch ID in the endpoint? So something like **GET /transfers/:batchid**?

I'm also thinking that we will need some kind of report on matrices. In particular, we'd like to be able to understand things like:
- Are there any batches which don't appear in a matrix?
- Are there any matrices which are in a hybrid state? I mean, some of their batches have been settled and some not, or some of their batches can be settled and some not.
- Unsettled matrices.

With regard to validation, @pedrosousabarreto, Paul Makin and I are the product owners for this workstream. I'd be happy to get Paul involved if you think it useful.

**Comment by pedrosousabarreto** [comment]
@MichaelJBRichards As product owner, can you create a high level requirements doc for settlements? One which focuses on the use cases from the operator perspective and not technical implementation details.

That would be great!

**Comment by MichaelJBRichards** [comment]
As requested, @pedrosousabarreto, I've created a requirements document for administrator functions relating to settlement. It's been reviewed by Paul Makin.

I want to reiterate a point which keeps cropping up. There should not be a requirement for the hub operator to take some action to update the content of a matrix. Updating a matrix is something that the system should take care of. This does not, of course, affect the idempotency of a request to see a settlement report, which should always return the current state of the report content. Leaving this responsibility with the hub operator is asking for operational trouble, in my view.

[Settlement administrator stories.docx](https://github.com/mojaloop/mojaloop-specification/files/11048711/Settlement.administrator.stories.docx)

If anyone has any immediate questions or comments, please reply; otherwise, we can talk about this at our next meeting.

**Comment by pedrosousabarreto** [comment]
Hi @MichaelJBRichards and Paul, thanks for this. We'll take proper look and provide detailed feedback as soon as possible.

From a first look, one issue comes out as something that was already discussed and explained a few times. We cannot fully re-calculate the matrix or the batches every time a new transfer is detected by the settlements system. As mentioned, this part of the system is on the high performance and high frequency path and would be too taxing on the system. We cannot build it like that. 

The way the system works, is that:
- Upon the arrival of a transfer committed event, the settlements system associates it to a batch, creating one if necessary. Any further processing at this stage will affect the ability of the system to scale.
- Later on, the operator can request a matrix to be created and recalculated. Because this request has an infinitely smaller rate of occurrence than a transfer, it will not affect the ability of the system to scale.

Another problem of the automatic matrix recalculation is that more transfers can be included in a matrix between the time the operator clicks the button to settle a matrix and the actual setting - because there are automatic processes happening in the background - this would result in a matrix that is different from the one the operator wanted to settle.

Think about it this way, why recalculate a matrix thousands or millions of times a day when we can do it once or twice without affecting user experience or functionality?

Why are you so adamant about this specific case?
I'm sure an operator wouldn't mind pressing a button to recalculate the matrix once or twice prior to settling it. This would make our system much easier on computational resources and cheaper to run.

**Comment by MichaelJBRichards** [comment]
I'm afraid I'm now confused, @pedrosousabarreto. First you tell me to leave the implementation details aside, and now you tell me that I have to modify my requirements because of the implementation details. Which path should I take?

**Comment by pedrosousabarreto** [comment]
The one were you don't ignore the fact that the automatic matrix recalculation on every transfer keeps us from delivering on the the non-functional requirements of scalability and performance. This was explained already several times, not sure what the confusion is.

**Comment by MichaelJBRichards** [comment]
The confusion is that I tried to explain why I don't accept that what you present (not explain) as a fact really is a fact and not a consequence of particular design decisions which could (and, in my opinion, should) be changed. You ignored my explanation and told me not to think about the implementation details. So I stopped. Now you tell me that I have to change my user requirements because of a "fact" for which no supporting evidence has been produced, and where alternative implementations have been proposed which do not appear to have the drawbacks of the current approach. Is that a clear enough statement of where the confusion lies?

**Comment by pedrosousabarreto** [comment]
I didn't ignore it Michael, would never do that.

You're proposing that we re-calculate batches and matrices on the fly when transfers are received, and not when an operator requests it to be recalculated. 
I'm saying this is a bad design. It is a performance problem that will require more hardware and keep us from scaling.
Given the minute different in UX for an operator I don't understand why this is a problem.

Am I understanding your request incorrectly?

**Comment by MichaelJBRichards** [comment]
My understanding was that we were already keeping the batch totals up to date, since the batch balance would be adjusted when a transfer was added to the settlement batch. If that isn't true, then I think we have a more extensive problem; but obviously we should confirm. So, if that's the case, this is simply a question about the summation of batches into matrices. My view -supported, I think, by experimental evidence - is that the summation of batches is unlikely to be a problem if we action it in response to a GET; but I'd certainly be interested in any evidence to the contrary.

So what I'm proposing is that we recalculate matrices, not batches, on demand, and that this is a problem small enough to be manageable in the context of a user interface. I repeat my view that, if we rely on users to request matrix updates and divorce those requests from viewing a matrix, people will wind up making mistakes because they will make the wrong assumptions about what they're looking at.

**Comment by MichaelJBRichards** [comment]
![image](https://user-images.githubusercontent.com/47815538/227297723-6a8009c4-9f26-4b96-bfaf-098086bbe8fb.png)
 
So when the reference architecture says (if I've understood the diagram correctly): "on receiving a _TransferCommittedFulfilled_ event, Settlements BC will create a settlement journal entry in the appropriate batch", that's not happening; or, at least, it's not happening in a way that would allow a "Get list of settlement batch account's balance" to trigger a simple "filter and return" action in Transfers BC?

I also don't see where the Settlements BC handles an event that triggers the recalculation of a batch. Did I miss something?

And I'm afraid I'm still not understanding the argument from scale. In my mind, the opposition is not between doing things thousands of times and doing them only once. Each of those transfers will eventually have to be processed exactly once in either mode; the question is whether you do them transfer by transfer, when you've got plenty of time and you can let things smooth out, or postpone them to a big bang when someone's waiting for the results. I had thought that, when Adrian and I modelled the Settlements BC for the reference architecture, we had opted for the first of those.

As to the question of the UI: I'm sort of happy if the Settlements Admin API contains a Get a Matrix resource which triggers calculation of the matrix and then returns it. I continue, though, to believe that separating them as resources at the API level will lead to operational and development mistakes.

**Comment by pedrosousabarreto** [comment]
No, that's the old version from 2021. There is a new version next to it from 2023. Following the meeting of the 8th I wrote in slack the capture requirements, did the demo at the convening and explained it, and in this issue I've explained several times how it is implemented currently in vNext settlements. As mentioned the only difference is the moment when the calculation happens and it has no impact on operator.

One last time: 
The current implementation, for valid reasons of performance, scalability and cost efficiency, does not calculate anything when receiving each transfer, at that time it only allocates the transfer to a batch. All calculations are performed only when an operator creates a matrix or requests its recalculation.

Anyway, I'm going to stop commenting here. I've commented here several times how it works, we're going in circles now.

I think we're delivering on all requirements, If this implementation is not acceptable, let's meet and discuss.

**Comment by elnyry-sam-k** [comment]
Following the conversation along, I felt that it'll be useful to have a final state for batches as well

Following what Jason @koekiebox as earlier
> • Change OPEN to CLOSED
> • Change OPEN to IN_DISPUTE
> • Change CLOSED to IN_DISPUTE
> • Change IN_DISPUTE to CLOSED

The last one can possibly be (or even after the initial move to 'closed', there maybe an element of timeline, similar to the step I've mentioned below)
> • Change IN_DISPUTE to SETTLED (a new term to indicate finality)

and there maybe a mechanism introduced around timing (for example, after a certain time disputes cannot be made for transfers)
> • Change OPEN to CLOSED initially and then after the scheme specified time,
> • Change CLOSED to SETTLED

The reason being that with the current structure because a CLOSED state doesn't mean finality, it leaves open the possibility to move a batch into dispute at any point, in theory.

**Comment by MichaelJBRichards** [comment]
Hi @elnyry-sam-k. With reference to your comments: a status of SETTLED is already proposed as part of the requirements (see the document further up this thread); but we should not allow batches to be changed to that status manually. Moving to a SETTLED status, implies changes to the position and settlement accounts for the selected settlement model, and should therefore only happen as part of the settlement process. The idea is that, when a batch comes out of the IN_DISPUTE status, it should have a status which allows it to be subsequently settled. It would, of course be open to the builders of an administrative UI to conflate the two processes, and provide functionality which first changed the status of the batch and subsequently settled it. In general, though, I'm happy with the idea that it should be possible to raise a dispute about a transfer at any point before the batch to which the transfer belongs is settled (i.e. while the status of the batch is OPEN or CLOSED.)

I would be against introducing a timing-based mechanism for changing status - at least, internally. We've generally tried to stick to an event-based syntax for the reference architecture, and that seems to me to be the right thing in this case. Again, it's always open to the builders of administrative functions to implement a Cron job to close batches or execute settlements based on the particular characteristics of their implementation, should they wish.

**Comment by koekiebox** [comment]
Thank you for the response @MichaelJBRichards. I hope that settles all @elnyry-sam-k ? 
Apologies, been hard at work on Settlements disputes the last couple of days.
